core/src/main/scala/org/bitcoins/core/wallet/signer/Signer.scala

```diff
@@ -74,55 +63,57 @@ sealed abstract class Signer {
-    *
-    * If you are signing a [[LockTimeScriptPubKey locktime spk]], you must sign
-    * the entire [[LockTimeScriptPubKey]], not the nested script.
-    */
-  def sigComponentToSignOpt: Option[TxSigComponent]
 
-  /** This represents the [[ScriptPubKey scriptpubkey]] for which a valid ScriptSignature is
-    * being generated. This is usually some nested SPK such as a redeem script in P2SH and P2WSH.
-    */
-  def scriptPubKeyToSatisfyOpt: Option[ScriptPubKey]
-}
+  protected val flags: Seq[ScriptFlag] = Policy.standardFlags
 
-object NestedSigning {
+  protected def relevantInfo(
+      spendingInfo: UTXOSpendingInfo,
+      unsignedTx: Transaction): (Seq[Sign], TransactionOutput, UInt32, HashType) = {
+    (spendingInfo.signers,
+     spendingInfo.output,
+     inputIndex(spendingInfo, unsignedTx),
+     spendingInfo.hashType)
+  }
 
-  /** The NoOverrides case where a Signer is used raw, without nesting */
-  case object NoOverrides extends OverridesForNestedSigning {
-    override val sigComponentToSignOpt: Option[TxSigComponent] = None
-    override val scriptPubKeyToSatisfyOpt: Option[ScriptPubKey] = None
+  protected def inputIndex(
+      spendingInfo: UTXOSpendingInfo,
+      tx: Transaction): UInt32 = {
+    tx.inputs.zipWithIndex
+      .find(_._1.previousOutput == spendingInfo.outPoint) match {
+      case Some((_, index)) => UInt32(index)
+      case None =>
+        throw new IllegalArgumentException(
+          "Transaction did not contain expected input.")
+    }
   }
 
-  /** For P2WSH signing, this will be passed to the nested Signer */
-  case class P2WSHOverrides(
-      externalSigComponent: TxSigComponent,
-      redeemScript: ScriptPubKey)
-      extends OverridesForNestedSigning {
-    override val sigComponentToSignOpt: Option[TxSigComponent] = Some(
-      externalSigComponent)
-    override val scriptPubKeyToSatisfyOpt: Option[ScriptPubKey] = Some(
-      redeemScript)
+  protected def sigComponent(
+      spendingInfo: UTXOSpendingInfo,
+      unsignedTx: Transaction): TxSigComponent = {
+    val index = inputIndex(spendingInfo, unsignedTx)
+
+    spendingInfo.output.scriptPubKey match {
+      case _: WitnessScriptPubKey =>
+        val wtx = unsignedTx match {
+          case btx: BaseTransaction =>
+            val transactionWitnessOpt =
+              spendingInfo.scriptWitnessOpt.map(scriptWit =>
+                TransactionWitness(Vector(scriptWit)))
+            val transactionWitness =
+              transactionWitnessOpt.getOrElse(EmptyWitness)
+
+            WitnessTransaction(btx.version,
+                               btx.inputs,
+                               btx.outputs,
+                               btx.lockTime,
+                               transactionWitness)
+          case wtx: WitnessTransaction => wtx
+        }
+
+        WitnessTxSigComponent(wtx, index, spendingInfo.output, flags)
+      case _: ScriptPubKey =>
```

Comment:
I don't think this accounts for the case of `P2SH(P2WSH())` does it? Shoouldn't that be a `WitnessTxSigComponentP2SH`?